metricbeat/module/jolokia/jmx/config.go

```diff
@@ -103,41 +104,258 @@ func (m AttributeMapping) Get(mbean, attr string) (Attribute, bool) {
 
-	// Using this regexp instead of just splitting by commas because values can be quoted
-	// and contain commas, what complicates the parsing.
-	properties := propertyRegexp.FindAllString(parts[1], -1)
-	propertyList := strings.Join(properties, ",")
+	// Create a new MBean object
+	mybean := &MBeanName{
+		Domain: parts[0],
+	}
+
+	// First of all verify that all bean properties are
+	// in the form key=value
+	tmpProps := propertyRegexp.FindAllString(parts[1], -1)
+	propertyList := strings.Join(tmpProps, ",")
 	if len(propertyList) != len(parts[1]) {
 		// Some property didn't match
-		return name, fmt.Errorf("mbean properties must be in the form key=value: %s", name)
+		return nil, fmt.Errorf("mbean properties must be in the form key=value: %s", mBeanName)
+	}
+
+	var mbeanRegexp = regexp.MustCompile("([^,=:*?]+)=([^,=:\"]+|\".*\")")
+
+	// Using this regexp we will split the properties in a 2 dimensional array
+	// instead of just splitting by commas because values can be quoted
+	// and contain commas, what complicates the parsing.
+	// For example this MBean property string:
+	//
+	// name=HttpRequest1,type=RequestProcessor,worker="http-nio-8080"
+	//
+	// will become:
+	//
+	// [][]string{
+	// 	[]string{"name=HttpRequest1", "name", "HttpRequest1"},
+	// 	[]string{"type=RequestProcessor", "type", "RequestProcessor"},
+	// 	[]string{"worker=\"http-nio-8080\"", "worker", "\"http-nio-8080\""}
+	// }
+	properties := mbeanRegexp.FindAllStringSubmatch(parts[1], -1)
+
+	// If we could not parse MBean properties
+	if properties == nil {
+		return nil, fmt.Errorf("mbean properties must be in the form key=value: %s", mBeanName)
+	}
+
+	// Initialise properties map
+	mybean.Properties = make(map[string]string)
+
+	for _, prop := range properties {
+
+		// If every row does not have 3 columns, then
+		// parsing must have failed.
+		if (prop == nil) || (len(prop) < 3) {
+			// Some property didn't match
+			return nil, fmt.Errorf("mbean properties must be in the form key=value: %s", mBeanName)
+		}
+
+		mybean.Properties[prop[1]] = prop[2]
+	}
+
+	return mybean, nil
+}
+
+// JolokiaHTTPRequest is a small struct which contains all request information
+// needed to construct a reqest helper.HTTP object which will be sent to Jolokia.
+// It is just an intermediary structure which can be easily tested as helper.HTTP
+// fields are all private.
+type JolokiaHTTPRequest struct {
+	// HttpMethod can be either "GET" or "POST"
+	HTTPMethod string
+	// URI which will be used to query Jolokia
+	URI string
+	// Request body which is only filled if the http method is "POST"
+	Body []byte
+}
+
+// JolokiaHTTPRequestBuilder is an interface which describes
+// the behaviour of the builder which generates the HTTP request
+// which is sent to Jolokia
+type JolokiaHTTPRequestBuilder interface {
+	// Fetches the information from Jolokia server regarding MBeans
+	BuildRequestsAndMappings(configMappings []JMXMapping) ([]*JolokiaHTTPRequest, AttributeMapping, error)
```

Comment:
As we not only need to customize the request but also the response handling depending on the method, maybe this Builder could abstract the whole `Fetch`.